File: fathersDay/Dsa/Day12 (224. Basic Calculator)/Solution.py
class Solution:
    def BasicCalculator(self, s: str) -> int:
        def calculate(s):
            stack = []
            num = 0
            sign = 1
            res = 0
            i = 0
            while i < len(s):
                char = s[i]
                if char.isdigit():
                    num = num * 10 + int(char)
                elif char == '+':
                    res += sign * num
                    num = 0
                    sign = 1
                elif char == '-':
                    res += sign * num
                    num = 0
                    sign = -1
                elif char == '(':
                    stack.append(res)
                    stack.append(sign)
                    res = 0
                    sign = 1
                elif char == ')':
                    res += sign * num
                    num = 0
                    res *= stack.pop()
                    res += stack.pop()
                i += 1
            res += sign * num
            return res

        return calculate(s.replace(' ', ''))

File: fathersDay/Dsa/Day12 (224. Basic Calculator)/Solution.java
import java.util.Stack;

class Solution {
    public int calculate(String s) {
        s = s.replace(" ", ""); // Remove spaces
        Stack<Integer> stack = new Stack<>();
        long num = 0;
        int sign = 1; // 1 for positive, -1 for negative
        int result = 0;
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (Character.isDigit(c)) {
                num = num * 10 + (c - '0');
            } else if (c == '+') {
                result += sign * num;
                num = 0;
                sign = 1;
            } else if (c == '-') {
                result += sign * num;
                num = 0;
                sign = -1;
            } else if (c == '(') {
                stack.push(result);
                stack.push(sign);
                result = 0;
                sign = 1;
            } else if (c == ')') {
                result += sign * num;
                num = 0;
                result *= stack.pop(); // Pop sign
                result += stack.pop(); // Pop previous result
            }
        }
        result += sign * num;
        return result;
    }
}